packages/flutter/lib/src/widgets/animated_raster.dart

```diff
@@ -0,0 +1,215 @@
+  /// const constructor so that subclasses can be const.
+  const AnimatedRasterDelegate();
+
+  /// Whether [paint] will paint anything on the current frame.
+  ///
+  /// This is used to optimize painting and delay rasterization if a child
+  /// would not be painted. One example would be an animated opacity, if the
+  /// first frame is fully transparent then rasterization can be delayed until
+  /// the second frame.
+  @protected
+  bool willPaint(Animation<double> animation) => true;
+
+  /// Paint the children of this render object.
+  ///
+  /// The provided [image] contains a GPU resident texture of the children of
+  /// this render object.
+  ///
+  /// The [pixelRatio] is the ratio of pixels in [image] to logical pixels.
+  @protected
+  void paint(PaintingContext context, ui.Image image, double pixelRatio, Animation<double> animation);
+}
+
+/// A widget that replaces [child] with a rasterized version while an animation is active.
+///
+/// How that image is painted is controlled by providing an [AnimatedRasterDelegate], which
+/// can also incorporate the animation value. This is primarily used for efficient page
+/// transitions.
+///
+/// This class depends on [Scene.toGpuImage] which is not supported on the html backend.
+///
+/// See also:
+///  * [ZoomPageTransitionsBuilder], which uses this render object to implement the fade and
+///    stretch effect efficiently.
+class AnimatedRaster extends SingleChildRenderObjectWidget {
+  /// Create a new [AnimatedRaster].
+  const AnimatedRaster({
+    super.key,
+    super.child,
+    required this.animation,
+    required this.delegate,
+  });
+
+  /// The animation that drives the [delegate].
+  final Animation<double> animation;
+
+  /// The delegate which controls how the child raster is painted.
+  final AnimatedRasterDelegate delegate;
+
+  @override
+  RenderObject createRenderObject(BuildContext context) {
+    return RenderAnimatedRaster(animation, delegate, MediaQuery.maybeOf(context)?.devicePixelRatio ?? 1.0);
+  }
+
+  @override
+  void updateRenderObject(BuildContext context, covariant RenderAnimatedRaster renderObject) {
+    renderObject
+      ..animation = animation
+      ..delegate = delegate
+      ..devicePixelRatio = MediaQuery.maybeOf(context)?.devicePixelRatio ?? 1.0;
+  }
+}
+
+/// A render object that converts its children to a [ui.Image] while animating.
+///
+/// While the status of [animation] is [AnimationStatus.completed] or
+/// [AnimationStatus.dismissed], children are painted normally. When the status
+/// is [AnimationStatus.forward] or [AnimationStatus.reverse], this render object
+/// will create a GPU resident texture the first time that [AnimatedRasterDelegate.willPaint] returns true
+/// and provide it to [AnimatedRasterDelegate.paint].
+///
+/// See also:
+///  * [ZoomPageTransitionsBuilder], which uses this render object to implement the fade and
+///    stretch effect efficiently.
+class RenderAnimatedRaster extends RenderProxyBox {
```

Comment:
Can/should this render object be private?

Reply:
_maybe_ 

I like the idea of a general purpose widget that we can point folks towards that makes toGpuImage easier to use. Whether or not this API is general purpose enough remains to be seen of course